pyomo/contrib/mindtpy/algorithm_base_class.py

```diff
@@ -1626,73 +1527,67 @@ def set_up_lazy_OA_callback(self, mainopt):
-            lazyoa = mainopt._solver_model.register_callback(
+            lazyoa = self.mip_opt._solver_model.register_callback(
                 single_tree.LazyOACallback_cplex
             )
             # pass necessary data and parameters to lazyoa
             lazyoa.main_mip = self.mip
-            self.solve_data = MindtPySolveData()
-            self.export_attributes()
-            lazyoa.solve_data = self.solve_data
             lazyoa.config = self.config
-            lazyoa.opt = mainopt
-            mainopt._solver_model.set_warning_stream(None)
-            mainopt._solver_model.set_log_stream(None)
-            mainopt._solver_model.set_error_stream(None)
+            lazyoa.opt = self.mip_opt
+            lazyoa.mindtpy_object = self
+            self.mip_opt._solver_model.set_warning_stream(None)
+            self.mip_opt._solver_model.set_log_stream(None)
+            self.mip_opt._solver_model.set_error_stream(None)
         if self.config.mip_solver == 'gurobi_persistent':
-            mainopt.set_callback(single_tree.LazyOACallback_gurobi)
+            self.mip_opt.set_callback(single_tree.LazyOACallback_gurobi)
+            self.mip_opt.mindtpy_object = self
+            self.mip_opt.config = self.config
 
     ##########################################################################################################################################
     # mip_solve.py
 
-    def solve_main(self, config):
+    def solve_main(self):
         """This function solves the MIP main problem.
 
-        Parameters
-        ----------
-        config : ConfigBlock
-            The specific configurations for MindtPy.
-
         Returns
         -------
         self.mip : Pyomo model
             The MIP stored in self.
         main_mip_results : SolverResults
             Results from solving the main MIP.
         """
+        config = self.config
         self.mip_iter += 1
 
         # setup main problem
-        self.setup_main(config)
-        mainopt, mip_args = self.set_up_mip_solver(config)
+        self.setup_main()
+        mip_args = self.set_up_mip_solver()
 
         try:
-            main_mip_results = mainopt.solve(
+            main_mip_results = self.mip_opt.solve(
                 self.mip, tee=config.mip_solver_tee, load_solutions=False, **mip_args
             )
             # update_attributes should be before load_from(main_mip_results), since load_from(main_mip_results) may fail.
-            if config.single_tree or config.use_tabu_list:
-                self.update_attributes()
             if len(main_mip_results.solution) > 0:
                 self.mip.solutions.load_from(main_mip_results)
-        except (ValueError, AttributeError):
+        except (ValueError, AttributeError, RuntimeError):
             if config.single_tree:
                 config.logger.warning('Single tree terminate.')
                 if get_main_elapsed_time(self.timing) >= config.time_limit - 2:
```

Comment:
The `timelimit` passed to MIP solver and NLP solvers should be `int`. Therefore, we use the following code to calculate the remaining time and pass it to MIP/NLP solvers.
```python
remaining = int(max(config.time_limit - elapsed, 1))
```
For example, given that
1. The `timelimit` passed to `MindtPy` is `100` seconds
2. `50.1` seconds have elapsed.
3. The `timelimit` passed to the next MIP/NLP solver is `int(49.9)`, which equals `49`.

The total elapsed time is `49 + 50.1 = 99.1` seconds, less than `100`. However, the algorithm terminates since it reaches the `timelimit`. In this case, we should set the `termination_condition` to `maxTimeLimit`.

That's why we use **subtract 2** here.

Reply:
Couldn't you just always take the ceiling of the remaining time to pass as `timelimit`? And then you might go a little over the specified time limit, but you won't have to have this correction factor to figure out what should have been a termination due to `maxTimeLimit`?

Reply:
I agree with Emma. This 2 seemed too out of the pocket.